### PR TITLE
[Bugfix] fix apply_temperature to avoid nan in probs

### LINKED ISSUE #22180: [Bug]: apply_temperature may cause nan in probs
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Your output of `python collect_env.py` here
```

</details>

### 🐛 Describe the bug

Recently, we met some illegal memory access problems. When we turned off flashinfer sampling, the problem disappeared, or at least it was much harder to trigger. Unfortunately, I can't provide a minimal reproduction here.

After some investigation, I believe the problem is caused by `apply_temperature`. Suppose we have a decode batch with both random sampling and greedy sampling. [Here](https://github.com/vllm-project/vllm/blob/a7b8788d2c2fae6bf52c128916de19e85f2b0a25/vllm/v1/worker/gpu_input_batch.py#L306) will set the temperature for greedy sampling to -1.0, and [here](https://github.com/vllm-project/vllm/blob/a7b8788d2c2fae6bf52c128916de19e85f2b0a25/vllm/v1/sample/sampler.py#L99) will div all the logits against temperature. If the logits (for greedy sampling) contains `-inf`, it will become `inf` after the division, and hence the probs tensor will be fully `nan`, I guess maybe it will cause bad things when using flashinfer top_p sampling. I'm not 100% sure about this, because I use probs with `nan` to run against flashinfer top_p sampling, and compute-sanitizer can't find any problems. But anyway, `nan` in probs is dangerous, I think it should be fixed.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by courage17340**
## Purpose
* This PR fixes https://github.com/vllm-project/vllm/issues/22180, and can solve many cases of illegal memory access when using flashinfer sampling.
* Just treat greedy temperature (0.0 now instead of -1.0) as 1.0 when performing apply_temperature is enough.

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by courage17340** [comment]
@njhill  Hi,  https://github.com/vllm-project/vllm/pull/22261 was not updated for a long time, so I proposed my PR here. Please take a look.

File: vllm/v1/sample/sampler.py [hunk 1]
```diff
@@ -131,7 +131,10 @@ def apply_temperature(
         temp: torch.Tensor,
     ) -> torch.Tensor:
         # Use in-place division to avoid creating a new tensor.
-        return logits.div_(temp.unsqueeze(dim=1))
+        # Avoid div by -1 (which is mask for greedy), because there may be
+        # -inf in logits, which can cause nan in probs.
+        return logits.div_(
+            torch.where(temp < _SAMPLING_EPS, 1.0, temp).unsqueeze(dim=1))
```
Comment (njhill) [inline vllm/v1/sample/sampler.py]: Thanks @courage17340. If we're doing this, we should have another "persistent" tensor in SamplingMetadata so that we don't have to create it here every time. 

Basically we would have two temp tensors, one used here, the other just used to keep track of which reqs are greedy vs non-greedy (which could then be bool-valued).

Comment (courage17340) [reply vllm/v1/sample/sampler.py]: OK, added greedy_mask and updated test

**Comment by njhill** [comment]
@courage17340 are you confident that this eliminates the IMA occurrences? We still can't reliably repro / don't understand the cause for sure, is that correct?

**Comment by njhill** [comment]
cc @WoosukKwon

**Comment by courage17340** [comment]
> @courage17340 are you confident that this eliminates the IMA occurrences? We still can't reliably repro / don't understand the cause for sure, is that correct?

I'm not 100% confident, but I think we can give it a try and see if it can eliminate IMA for others too.

https://github.com/flashinfer-ai/flashinfer/issues/1575 This issue says nan in probs can cause IMA in flashinfer sampling.

**Comment by KingsleyZhang123** [comment]
Thanks for the fix!
We have also seen cuda IMA triggered by FlashInfer sampling recently, and after applying this fix, the issue has gone.
cc: @houseroad

**Comment by njhill** [comment]
@courage17340 I discussed this with @WoosukKwon, he would prefer a simpler solution without an additional persistent tensor: just do this for the temperature op:
```python
        temp = torch.where(temp > _SAMPLING_EPS, temp, 1.0)
        return logits.div_(temp.unsqueeze(dim=1))
```

Then we can also just keep greedy entries as 0 rather than -1. I guess we could still skip the extra torch.where in the all_random case.

**Comment by courage17340** [comment]
> @courage17340 I discussed this with @WoosukKwon, he would prefer a simpler solution without an additional persistent tensor: just do this for the temperature op:
> 
> ```python
>         temp = torch.where(temp > _SAMPLING_EPS, temp, 1.0)
>         return logits.div_(temp.unsqueeze(dim=1))
> ```
> 
> Then we can also just keep greedy entries as 0 rather than -1. I guess we could still skip the extra torch.where in the all_random case.

OK, changed back to the first verison and skipped torch.where when all_random.

**Comment by njhill** [comment]
Thanks @courage17340, I think we can also remove the -1 stuff in `input_batch_gpu.py` too to avoid future confusion (just keep greedy as 0), and then no need to mention in the comment.. can just say it is to avoid div by 0.

**Comment by courage17340** [comment]
> Thanks @courage17340, I think we can also remove the -1 stuff in `input_batch_gpu.py` too to avoid future confusion (just keep greedy as 0), and then no need to mention in the comment.. can just say it is to avoid div by 0.

Done.

**Comment by courage17340** [comment]
@njhill  Hello, any more comments? Can this be merged now?

**Review by njhill — APPROVED** [review]
Thanks @courage17340
